src/WebJobs.Script.WebHost/Metrics/LegionMetricsFileManager.cs

```diff
@@ -0,0 +1,112 @@
+        private readonly ILogger _logger;
+
+        public LegionMetricsFileManager(string metricsFilePath, IFileSystem fileSystem, ILogger logger, int maxFileCount)
+        {
+            _fileSystem = fileSystem ?? throw new ArgumentNullException(nameof(fileSystem));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            MetricsFilePath = metricsFilePath;
+            _maxFileCount = maxFileCount;
+        }
+
+        internal string MetricsFilePath { get; set;  }
+
+        private bool PrepareDirectoryForFile()
+        {
+            if (string.IsNullOrEmpty(MetricsFilePath))
+            {
+                return false;
+            }
+
+            // ensure the directory exists
+            _fileSystem.Directory.CreateDirectory(MetricsFilePath);
+
+            var metricsDirectoryInfo = _fileSystem.DirectoryInfo.FromDirectoryName(MetricsFilePath);
+            var files = metricsDirectoryInfo.GetFiles().OrderBy(p => p.CreationTime).ToList();
+
+            // ensure we're under the max file count
+            if (files.Count < _maxFileCount)
+            {
+                return true;
+            }
+
+            // we're at or over limit
+            // delete enough files that we have space to write a new one
+            int numToDelete = files.Count - _maxFileCount + 1;
+            var filesToDelete = files.Take(numToDelete).ToArray();
+
+            _logger.LogDebug($"Deleting {filesToDelete.Length} metrics file(s).");
+
+            Parallel.ForEach(filesToDelete, file =>
+            {
+                try
+                {
+                    file.Delete();
+                }
+                catch (Exception ex) when (!ex.IsFatal())
+                {
+                    // best effort
+                    _logger.LogError(ex, $"Error deleting metrics file '{file.FullName}'.");
+                }
+            });
+
+            files = metricsDirectoryInfo.GetFiles().OrderBy(p => p.CreationTime).ToList();
+
+            // return true if we have space for a new file
+            return files.Count < _maxFileCount;
+        }
+
+        public async Task PublishMetricsAsync(object metrics)
+        {
+            string fileName = string.Empty;
+
+            try
+            {
+                bool metricsPublishEnabled = !string.IsNullOrEmpty(MetricsFilePath);
+                if (metricsPublishEnabled && !PrepareDirectoryForFile())
+                {
+                    return;
+                }
+
+                var settings = new JsonSerializerSettings
+                {
+                    NullValueHandling = NullValueHandling.Ignore
+                };
+
+                string metricsContent = JsonConvert.SerializeObject(metrics, settings);
+                _logger.PublishingMetrics(metricsContent);
+
+                if (metricsPublishEnabled)
+                {
+                    fileName = $"{Guid.NewGuid().ToString().ToLower()}.json";
```

Comment:
Any particular reason this needs to be in lowercase?

```suggestion
                    fileName = $"{Guid.NewGuid().ToString().ToLowerInvariant()}.json";
```

(or remove the ToLower call altogether)

Reply:
Made the change you suggested. I can't remember now why I lower cased it, but I'll leave as is since we have Legion consuming these files.